internal/pkg/api/handleUpload.go

```diff
@@ -0,0 +1,390 @@
+	if err != nil {
+		return fmt.Errorf("error reading request: %w", err)
+	}
+
+	// decode once here to access known fields we need to parse and work with
+	var fi FileInfo
+	if err := json.Unmarshal(body, &fi); err != nil {
+		if errors.Is(err, io.EOF) {
+			return fmt.Errorf("file info body is required: %w", err)
+		}
+		return err
+	}
+
+	// check API key matches payload agent ID
+	if _, err := authAgent(r, &fi.AgentID, ut.bulker, ut.cache); err != nil {
+		return err
+	}
+
+	if err := validateUploadPayload(fi); err != nil {
+		return err
+	}
+
+	docID := fmt.Sprintf("%s.%s", fi.ActionID, fi.AgentID)
+
+	var hasher hash.Hash
+	var sum string
+	switch {
+	case fi.File.Hash.SHA256 != "":
+		hasher = sha256.New()
+		sum = fi.File.Hash.SHA256
+	case fi.File.Hash.MD5 != "":
+		hasher = md5.New()
+		sum = fi.File.Hash.MD5
+	}
+
+	op, err := ut.upl.Begin(fi.File.Size, docID, fi.Source, sum, hasher)
+	if err != nil {
+		return err
+	}
+
+	// second decode here to maintain the arbitrary shape and fields we will just pass through
+	var reqDoc map[string]interface{}
+	if err := json.Unmarshal(body, &reqDoc); err != nil {
+		return fmt.Errorf("error parsing request json: %w", err)
+	}
+
+	doc, err := uploadRequestToFileDoc(reqDoc, op.ChunkSize)
+	if err != nil {
+		return fmt.Errorf("unable to convert request to file metadata doc: %w", err)
+	}
+	ret, err := upload.CreateFileDoc(r.Context(), ut.bulker, doc, fi.Source, docID)
+	if err != nil {
+		return err
+	}
+
+	zlog.Info().Str("return", ret).Msg("wrote doc")
+
+	out, err := json.Marshal(map[string]interface{}{
+		"upload_id":  op.ID,
+		"chunk_size": op.ChunkSize,
+	})
+	if err != nil {
+		return err
+	}
+	w.Header().Set("Content-Type", "application/json")
+	_, err = w.Write(out)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ut *UploadT) handleUploadChunk(zlog *zerolog.Logger, w http.ResponseWriter, r *http.Request, uplID string, chunkID int) error {
+	chunkInfo, err := ut.upl.Chunk(uplID, chunkID)
+	if err != nil {
+		return err
+	}
+	defer chunkInfo.Token.Release()
+	if chunkInfo.FirstReceived {
+		if err := updateUploadStatus(r.Context(), ut.bulker, chunkInfo.Upload, UploadProgress); err != nil {
+			zlog.Warn().Err(err).Str("upload", uplID).Msg("unable to update upload status")
+		}
+	}
+
+	// prevent over-sized chunks
+	data := http.MaxBytesReader(w, r.Body, upload.MaxChunkSize)
+	ce := cbor.NewChunkWriter(data, chunkInfo.Final, chunkInfo.Upload.DocID, chunkInfo.Upload.ChunkSize)
+	if err := upload.IndexChunk(r.Context(), ut.chunkClient, ce, chunkInfo.Upload.Source, chunkInfo.Upload.DocID, chunkInfo.ID); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ut *UploadT) handleUploadComplete(zlog *zerolog.Logger, w http.ResponseWriter, r *http.Request, uplID string) error {
+	info, err := ut.upl.Complete(uplID, ut.bulker)
+	if err != nil {
+		return err
+	}
+
+	if err := updateUploadStatus(r.Context(), ut.bulker, info, UploadDone); err != nil {
+		// should be 500 error probably?
```

Comment:
In this error case, the complete file has been uploaded and verified, but the status is stuck, a 500 would be safe, but it would be a very different error then if the `Complete` call above failed.